Added CSRF validation (closes #362, #375)

Koc opened the pull request:
Closes #362, #375 

TODO:
- [x] Add tests

Koc commented on Resources/config/csrf.xml:
```diff
@@ -0,0 +1,13 @@
+
+    <services>
+        <service id="sensio_framework_extra.security.listener" class="Sensio\Bundle\FrameworkExtraBundle\EventListener\CsrfListener">
+            <argument type="service" id="security.csrf.token_manager" on-invalid="null" />
```
Can I make it required and validate existence when creating new instance?

xabbuh replied:
Why not simply removing this definition if the `security.csrf.token_manager` service is not available?

Koc commented:
@fabpot tests pass. Please review

jakzal commented on EventListener/CsrfListener.php:
```diff
@@ -0,0 +1,61 @@
+        $token = $request->get($configuration->getParam());
+
+        if (!$this->csrfTokenManager->isTokenValid(new CsrfToken($configuration->getIntention(), $token))) {
+            throw new BadRequestHttpException('Csrf token is invalid.');
```
I found examples of both. [Spring returns a 403](http://docs.spring.io/spring-security/site/docs/current/reference/html/csrf.html#csrf-configure):

> The next step is to include Spring Security’s CSRF protection within your application. Some frameworks handle invalid CSRF tokens by invaliding the user’s session, but this causes its own problems. Instead by default Spring Security’s CSRF protection will produce an HTTP 403 access denied.

fabpot replied:
I think I would slightly prefer the AccessDenied one.

fabpot commented on Configuration/Csrf.php:
```diff
@@ -0,0 +1,62 @@
+{
+    protected $intention;
+
+    protected $param = '_token';
```
Can we make the properties private?

fabpot commented on Resources/doc/annotations/csrf.rst:
```diff
@@ -0,0 +1,38 @@
+@Security
```
`@Csrf`?

fabpot commented on Resources/doc/annotations/csrf.rst:
```diff
@@ -0,0 +1,38 @@
+
+.. caution::
+
+    The ``@Csrf`` annotation only works as of Symfony 2.7.
```
`The ``@Csrf`` annotation requires Symfony 2.7+.`?

fabpot commented on Resources/doc/annotations/csrf.rst:
```diff
@@ -0,0 +1,38 @@
+
+On the template side
+
+    <a href="{{ path('acme_post_delete', {'_token': csrf_token('post_delete')}) }}">delete post</a>
```
I would prefer that you use an example that does not use GET as it is really a bad practice that we should not encourage.

fabpot commented on Tests/EventListener/CsrfListenerTest.php:
```diff
@@ -0,0 +1,106 @@
+    {
+        if (!interface_exists('Symfony\Component\Security\Csrf\CsrfTokenManagerInterface')) {
+            $this->markTestSkipped('This test requires "symfony/security-csrf".');
+        }
```
This should be removed as we list all required deps in composer.json so it should always be present.

fabpot commented on Tests/EventListener/CsrfListenerTest.php:
```diff
@@ -0,0 +1,106 @@
+            ->getMockBuilder('Symfony\Component\Security\Csrf\CsrfTokenManagerInterface')
+            ->getMock();
+
+        $this->listener = new CsrfListener($this->tokenManager);
```
Please remove the `setUp()` method and the property and creates instances directly in tests. You can of course create helper methods like you did for the Request and the Event.

Koc commented:
update:
* annotation properties maked private
* added default intention equal to `default`
* updated docs
* updated tests
* created own exception. I doubt that reusing AccessDeniedException is good idea because of it depends on attributes and subject and it was created for authorization errors. Also this new exception allow pass valid token - useful for handling invalid tokens on client - we can refresh token, ask user and allow retry last action with new token.

Koc commented:
tests failures looks like unrelated to this PR

xabbuh commented on EventListener/CsrfListener.php:
```diff
@@ -0,0 +1,62 @@
+{
+    private $csrfTokenManager;
+
+    public function __construct(CsrfTokenManagerInterface $csrfTokenManager = null)
```
Allowing `null` here makes it possible to create the listener in an unusable state. I think you should rather remove (or not create) the definition instead when no CSRF token manager is available (see also my comment below).

Koc replied:
I've just reflect logic from security listener.

xabbuh commented on Resources/doc/annotations/csrf.rst:
```diff
@@ -0,0 +1,42 @@
+
+On the template side
+
+    <form action="{{ path('acme_posts_batch_delete') }}" method="POST">
```
This should be put into a code block too.

HeahDude replied:
You can use a double colon for PHP code blocks:
```rst
On the template side::

    <form action="{{ path('acme_posts_batch_delete') }}" method="POST">
```

HeahDude replied:
Edited my post since the code is html :)

Koc commented:
rebased on master again

ping @fabpot 

questions that required to be resolved:
1a. should I add and register custom exception handler for converting `InvalidCsrfTokenException` to 403 response? Invalid csrf produces 500 error for now. 
1b. AccessDeniedException is not good idea because reasons described in https://github.com/sensiolabs/SensioFrameworkExtraBundle/pull/430#issuecomment-249187438
1c. Or extends from `HttpException` from http-kernel?
2. my listener reflects logic from security listener. But is it possible check availability of the required services inside configuration and enable/disable required annotations depends on services availability. Like https://github.com/symfony/symfony/blob/master/src/Symfony/Bundle/FrameworkBundle/DependencyInjection/Configuration.php#L512 but we need check services.

Koc commented:
@fabpot can you make review and give advices to my previous questions, please? Would be nice finish this PR.

fabpot commented:
Closing as this repository is not maintained anymore. As of Symfony 6.2, all features have been moved to Symfony core now.
See #783
